test/parallel/test-fs-read-position-validation.mjs

```diff
@@ -14,73 +14,26 @@ const length = buffer.byteLength;
 // allowedErrors is an array of acceptable internal errors
 // For example, on some platforms read syscall might return -EFBIG
 async function testValid(position, allowedErrors = []) {
-  let fdSync;
-  try {
-    fdSync = fs.openSync(filepath, 'r');
-    fs.readSync(fdSync, buffer, offset, length, position);
-    fs.readSync(fdSync, buffer, { offset, length, position });
-  } catch (err) {
-    if (!allowedErrors.includes(err.code)) {
-      assert.fail(err);
-    }
-  } finally {
-    if (fdSync) fs.closeSync(fdSync);
-  }
-
-  fs.open(filepath, 'r', common.mustSucceed((fd) => {
-    try {
-      if (allowedErrors.length) {
-        fs.read(fd, buffer, offset, length, position, common.mustCall((err, ...results) => {
-          if (err && !allowedErrors.includes(err.code)) {
-            assert.fail(err);
-          }
-        }));
-        fs.read(fd, { buffer, offset, length, position }, common.mustCall((err, ...results) => {
-          if (err && !allowedErrors.includes(err.code)) {
-            assert.fail(err);
-          }
-        }));
-      } else {
-        fs.read(fd, buffer, offset, length, position, common.mustSucceed());
-        fs.read(fd, { buffer, offset, length, position }, common.mustSucceed());
-      }
-    } finally {
-      fs.close(fd, common.mustSucceed());
-    }
-  }));
-}
-
-async function testInvalid(code, position, internalCatch = false) {
-  let fdSync;
-  try {
-    fdSync = fs.openSync(filepath, 'r');
-    assert.throws(
-      () => fs.readSync(fdSync, buffer, offset, length, position),
-      { code }
-    );
-    assert.throws(
-      () => fs.readSync(fdSync, buffer, { offset, length, position }),
-      { code }
-    );
-  } finally {
-    if (fdSync) fs.closeSync(fdSync);
-  }
-
-  // Set this flag for catching errors via first argument of callback function
-  if (internalCatch) {
+  return new Promise((resolve, reject) => {
     fs.open(filepath, 'r', common.mustSucceed((fd) => {
-      try {
-        fs.read(fd, buffer, offset, length, position, (err, ...results) => {
-          assert.strictEqual(err.code, code);
-        });
-        fs.read(fd, { buffer, offset, length, position }, (err, ...results) => {
-          assert.strictEqual(err.code, code);
-        });
-      } finally {
-        fs.close(fd, common.mustSucceed());
-      }
+      let callCount = 3;
+      const handler = common.mustCall((err) => {
+        callCount--;
+        if (err && !allowedErrors.includes(err.code)) {
+          fs.close(fd, common.mustSucceed(() => reject(err)));
```

Comment:
We should reject the promise straight away, otherwise another read operation may resolve it before the file closes, and the test would not be reported as failure.
```suggestion
          fs.close(fd, common.mustSucceed());
          reject(err);
```